### PR TITLE
feat(TlsSocket): add ownership semantics to TlsSocket

### PR DESCRIPTION
**Pull request by kostasrim**
TlsSocket is constructed by a pointer to the next socket, effectively acting as a decorator/proxy object of the underline socket. This however is problematic, because once a connection upgrades to tls, you now need to keep alive two different objects with whose lifetime semantics are bound. Therefore, we make TlsSocket own the underline socket, to avoid decoupling the lifetime of the two sockets.

### CODE REVIEW

**Comment by kostasrim** [comment]
Do not merge this unless my other PR passes all tests. I will comment here :)

File: util/tls/tls_engine_test.cc
Comment (romange) [inline util/tls/tls_engine_test.cc:337]: why did you delete it?

Comment (kostasrim) [reply util/tls/tls_engine_test.cc:337]: because it defaults construct a `TlsSocket`, this test does nothing :D Also, TlsSocket, is no longer default constructable

File: util/fiber_socket_base.h
Comment (royjacobson) [inline util/fiber_socket_base.h:152]: formatting?

Comment (kostasrim) [reply util/fiber_socket_base.h:152]: That's really weird, 99.9% it's a clang-format bug, because I did run it before I pushed

Comment (kostasrim) [reply util/fiber_socket_base.h:152]: and speaking of that I will add pre-commit here.

**Review by romange — APPROVED** [review]

**Pull request merged**
